shared/desktop/app/main-window.desktop.tsx

```diff
@@ -32,75 +35,166 @@ export default function() {
-        ? appState.state.useNativeFrame
-        : defaultUseNativeFrame,
-    height: appState.state.height,
-    minHeight: 600,
-    minWidth: 400,
-    show: false,
-    webPreferences: {
-      devTools: showDevTools,
-      nodeIntegration: true,
-      nodeIntegrationInWorker: false,
-    },
-    width: appState.state.width,
-    x: appState.state.x,
-    y: appState.state.y,
-    ...(isDarwin ? {titleBarStyle: 'hiddenInset'} : {}),
+const defaultWindowState: WindowState = {
+  dockHidden: false,
+  height: 600,
+  isFullScreen: false,
+  width: 800,
+  windowHidden: false,
+  x: 0,
+  y: 0,
+}
+
+const windowState = {...defaultWindowState}
+
+const setupWindowEvents = (win: Electron.BrowserWindow) => {
+  const saveWindowState = debounce(() => {
+    let winBounds = win.getNormalBounds()
+    windowState.x = winBounds.x
+    windowState.y = winBounds.y
+    windowState.width = winBounds.width
+    windowState.height = winBounds.height
+    windowState.isFullScreen = win.isFullScreen()
+    windowState.windowHidden = !win.isVisible()
+    mainWindowDispatch(ConfigGen.createUpdateWindowState({windowState}))
+  }, 5000)
+
+  win.on('show', saveWindowState)
+  win.on('close', saveWindowState)
+  win.on('resize', saveWindowState)
+  win.on('move', saveWindowState)
+
+  const hideInsteadOfClose = (event: Electron.Event) => {
+    event.preventDefault()
+    win.hide()
+    hideDockIcon()
+  }
+
+  win.on('close', hideInsteadOfClose)
+
+  type IPCPayload = {type: 'showMainWindow'} | {type: 'closeWindows'}
+  Electron.app.on('KBkeybase' as any, (_: string, payload: IPCPayload) => {
+    switch (payload.type) {
+      case 'showMainWindow':
+        win.show()
+        showDockIcon()
+        break
+      case 'closeWindows':
+        hideDockIcon()
+        break
+    }
   })
+}
 
-  const webContents = mainWindow.window && mainWindow.window.webContents
+const changeDock = (show: boolean) => {
+  const dock = Electron.app.dock
+  if (!dock) return
 
-  if (showDevTools) {
-    webContents && webContents.openDevTools({mode: 'detach'})
+  if (show === dock.isVisible()) {
+    return
   }
 
-  appState.manageWindow(mainWindow.window)
+  if (show) {
+    dock.show()
+  } else {
+    dock.hide()
+  }
 
-  const app = SafeElectron.getApp()
-  // Register for SEP7 and Keybase links.
-  app.setAsDefaultProtocolClient('web+stellar')
-  app.setAsDefaultProtocolClient('keybase')
+  windowState.dockHidden = !show
+  mainWindowDispatch(ConfigGen.createUpdateWindowState({windowState}))
+}
+
+export const showDockIcon = () => changeDock(true)
+export const hideDockIcon = () => changeDock(false)
+
+let useNativeFrame = defaultUseNativeFrame
+
+const loadWindowState = () => {
+  const filename = dataRoot + 'gui_config.json'
 
-  const openedAtLogin = app.getLoginItemSettings().wasOpenedAtLogin
+  let openAtLogin = true
+  try {
+    const s = fs.readFileSync(filename, {encoding: 'utf8'})
```

Comment:
how will this work with sandboxing?
also, does this pr migrate your old setting?